// angular-clipboard-auto/index.js
import { Directive, ElementRef, NgModule, Renderer2 } from '@angular/core';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var CopyDirective = (function () {
    function CopyDirective(_el, _renderer) {
        this._el = _el;
        this._renderer = _renderer;
    }
    /**
     * @return {?}
     */
    CopyDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var /** @type {?} */ copy_icon = this._renderer.createElement('i');
        this._renderer.addClass(copy_icon, 'fa');
        this._renderer.addClass(copy_icon, 'fa-copy');
        this._renderer.addClass(copy_icon, 'angular-copy-icon');
        var /** @type {?} */ simple = this._renderer.listen(copy_icon, 'click', function (evt) {
            _this.copy(_this._el.nativeElement);
        });
        this._renderer.appendChild(this._el.nativeElement.parentNode, copy_icon);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    CopyDirective.prototype.copy = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var /** @type {?} */ input = event;
        input.select();
        document.execCommand('copy');
        input.setSelectionRange(-1, -1);
    };
    CopyDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[appCopy]'
                },] },
    ];
    /** @nocollapse */
    CopyDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: Renderer2, },
    ]; };
    return CopyDirective;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var ClipboardModule = (function () {
    function ClipboardModule() {
    }
    /**
     * @return {?}
     */
    ClipboardModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: ClipboardModule,
            providers: []
        };
    };
    ClipboardModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        CopyDirective
                    ],
                    exports: [
                        CopyDirective
                    ]
                },] },
    ];
    return ClipboardModule;
}());

export { ClipboardModule };
